File: classes/Zed.py
from Scalar import Scalar
from Complex import Complex

zero   = Scalar(0,0)	
origin = Complex( zero, zero )

class Zed:
	
	def __init__(self):
		self.position = origin.clone()
		self.iterator = self.position.clone()
		self.generation = 1
		self.escaped = False
	
	def refresh(self, position):
		self.position.copy(position)
		self.iterator.copy(self.position)
		self.generation = 1
		self.escaped = False
		self.hasEscaped()
	
	def iterate(self):
		if (not self.escaped):
			self.iterator.square()
			self.iterator.addBy(self.position)
			self.generation += 1
		return self.hasEscaped()
	
	def hasEscaped(self):
		if (not self.escaped):
			if (self.iterator.magnitude() > 4.0):
				print "escaped at generation", self.generation
				self.escaped = True
		return self.escaped

File: java/dburggie/mandelbrot/Zed.java
package dburggie.mandelbrot;

public class Zed
{
	private Complex position;
	private Complex iteration;
	private int generation;
	private boolean bound;
	
	public Zed()
	{
		this( new Complex(0.0,0.0) );
	}
	
	public Zed(Complex p)
	{
		position = p.clone();
		iteration = position.clone();
		generation = 1;
		bound = true;
		hasEscaped();
	}
	
	public void refresh(double x, double y)
	{
		position.setXY(x,y);
		iteration.copy(position);
		generation = 1;
		bound = true;
		hasEscaped();
	}
	
	public void refresh(Complex p)
	{
		position.copy(p);
		iteration.copy(position);
		generation = 1;
		bound = true;
		hasEscaped();
	}
	
	public void iterate(int n)
	{
		for (int i = 0; i < n; i++)
		{
			if (!bound)
			{
				break;
			}
			
			iteration.square();
			iteration.addBy(position);
			hasEscaped();
		}
	}
	
	public int hasEscaped()
	{
		if (bound)
		{
			if (iteration.magnitude() > 4.0)
			{
				bound = false;
				return generation;
			}
			
			return 0;
		}
		
		return generation;
	}
}
